File: output/GPT4o-Mini/xCodeEval_4000/C++/Go/b692a74ed0ad0a0c0ee36100c5b86f79.go
package main

import (
	"fmt"
	"math"
)

func main() {
	n := readInt()
	a := readInts(n)
	for i := range a {
		a[i]--
	}

	ans := 0
	fre := -1

	cnt := make(map[int]int)
	for _, e := range a {
		cnt[e]++
	}
	ma := 0
	for _, y := range cnt {
		if y > ma {
			ma = y
		}
	}
	mac := 0
	for x, y := range cnt {
		if y == ma {
			mac++
			fre = x
		}
	}
	if mac > 1 {
		ans = n
	}

	const maxv = 100
	for i := 0; i < maxv; i++ {
		if i == fre {
			continue
		}
		b := []int{0}
		for _, e := range a {
			pb := b[len(b)-1]
			if e == fre {
				pb--
			} else if e == i {
				pb++
			}
			b = append(b, pb)
		}
		mi := make([]int, 2*n+1)
		ma := make([]int, 2*n+1)
		for j := range mi {
			mi[j] = n + 1
			ma[j] = -1
		}
		for i := 0; i <= n; i++ {
			va := n + b[i]
			chle(&mi[va], i)
			chge(&ma[va], i)
		}
		j := n + 1
		for i := 0; i < len(mi); i++ {
			chle(&j, mi[i])
			chge(&ans, ma[i]-j)
		}
	}

	fmt.Println(ans)
}

func chle(x *int, y int) {
	if y < *x {
		*x = y
	}
}

func chge(x *int, y int) {
	if y > *x {
		*x = y
	}
}

func readInt() int {
	var x int
	fmt.Scan(&x)
	return x
}

func readInts(n int) []int {
	x := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&x[i])
	}
	return x
}

File: output/GPT4o-Mini/xCodeEval_remove_4000/C++/Python/b692a74ed0ad0a0c0ee36100c5b86f79.py
from collections import defaultdict
import sys

eol = '\n'

def read_int():
    return int(sys.stdin.readline().strip())

def read_ints():
    return list(map(int, sys.stdin.readline().strip().split()))

def main():
    n = read_int()
    a = read_ints()
    a = [x - 1 for x in a]

    ans = 0
    fre = -1

    cnt = defaultdict(int)
    for e in a:
        cnt[e] += 1

    ma = max(cnt.values())
    mac = sum(1 for y in cnt.values() if y == ma)
    fre = next(x for x, y in cnt.items() if y == ma) if mac == 1 else fre

    if mac > 1:
        ans = n

    maxv = 100
    for i in range(maxv):
        if i == fre:
            continue
        b = [0]
        for e in a:
            pb = b[-1]
            if e == fre:
                pb -= 1
            elif e == i:
                pb += 1
            b.append(pb)

        mi = [n + 1] * (2 * n + 1)
        ma = [-1] * (2 * n + 1)

        for i in range(n + 1):
            va = n + b[i]
            mi[va] = min(mi[va], i)
            ma[va] = max(ma[va], i)

        j = n + 1
        for i in range(len(mi)):
            j = min(j, mi[i])
            ans = max(ans, ma[i] - j)

    print(ans, end=eol)

if __name__ == "__main__":
    main()
